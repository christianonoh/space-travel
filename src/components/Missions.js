import { useSelector } from 'react-redux';
import styles from '../styles/missions.module.css';
import Button from './Button';

const Missions = () => {
  const missionsArr = useSelector((store) => store.missions);

  return (
    <section>
      <table className={styles['mission-table']}>
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
            <th>{' '}</th>
          </tr>
        </thead>
        <tbody>
          {missionsArr.map((row) => (
            <tr key={row.mission_id}>
              <td>{row.mission_name}</td>
              <td>{row.description}</td>
              <td>
                {row.status ? (
                  <span className={`${styles.badge} ${styles['badge-success']}`}>Member</span>
                ) : (
                  <span className={`${styles.badge} ${styles['badge-danger']}`}>Not a member</span>
                )}
              </td>
              <td>
                <Button title={row.status ? 'Leave' : 'Join'} />
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </section>
  );
};

export default Missions;
