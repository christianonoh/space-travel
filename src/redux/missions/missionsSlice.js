import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

export const fetchData = createAsyncThunk('missions/fetchData', async () => {
  const config = {
    method: 'get',
    maxBodyLength: Infinity,
    url: 'https://api.spacexdata.com/v3/missions/',
    headers: {},
  };
  const response = await axios.request(config);
  if (response.data) {
    return response.data;
  }
  return [];
});

const initialState = {
  data: [],
  status: 'idle',
  error: null,
};

const missionsReducer = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    toggleMissionReserved: (state, action) => {
      const missionId = action.payload;
      const newData = state.data.map((mission) => {
        if (mission.mission_id !== missionId) return mission;
        return { ...mission, reserved: !mission.reserved };
      });
      return { ...state, data: newData };
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchData.fulfilled, (state, action) => {
      const dataWithReserved = action.payload.map((mission) => ({
        ...mission,
        reserved: false,
      }));
      return { ...state, data: dataWithReserved, status: 'fulfilled' };
    });
  },
});

export const { toggleMissionReserved } = missionsReducer.actions;
export default missionsReducer.reducer;
