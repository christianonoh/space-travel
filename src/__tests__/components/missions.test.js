import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { render } from '@testing-library/react';
import Missions from '../../components/Missions';
import { fetchData, toggleMissionReserved } from '../../redux/missions/missionsSlice';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

jest.mock('../../redux/missions/missionsSlice', () => ({
  fetchData: jest.fn(),
  toggleMissionReserved: jest.fn(),
}));

describe('Missions component', () => {
  let mockMissionsSelector;
  let mockDispatch;

  beforeEach(() => {
    mockMissionsSelector = jest.fn();
    mockDispatch = jest.fn();
    useSelector.mockImplementation((selectorFn) => selectorFn({ missions: mockMissionsSelector() }));
    useDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    useSelector.mockClear();
    useDispatch.mockClear();
    fetchData.mockClear();
    toggleMissionReserved.mockClear();
  });

  it('renders correctly', () => {
    mockMissionsSelector.mockReturnValue({
      status: 'idle',
      data: [
        {
          mission_id: 1,
          mission_name: 'Mission 1',
          description: 'Mission 1 Description',
          reserved: true,
        },
        {
          mission_id: 2,
          mission_name: 'Mission 2',
          description: 'Mission 2 Description',
          reserved: false,
        },
      ],
    });

    const { container } = render(<Missions />);

    expect(container).toMatchSnapshot();
  });
});
