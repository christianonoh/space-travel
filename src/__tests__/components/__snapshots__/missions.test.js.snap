// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions component renders correctly 1`] = `
<div>
  <section>
    <table
      class="mission-table"
    >
      <thead>
        <tr>
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
          <th>
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Mission 1
          </td>
          <td>
            Mission 1 Description
          </td>
          <td>
            <span
              class="badge badge-success"
            >
              ACTIVE MEMBER
            </span>
          </td>
          <td>
            <button
              class="leaveBtn"
              type="button"
            >
              Leave Mission
            </button>
          </td>
        </tr>
        <tr>
          <td>
            Mission 2
          </td>
          <td>
            Mission 2 Description
          </td>
          <td>
            <span
              class="badge badge-danger"
            >
              NOT A MEMBER
            </span>
          </td>
          <td>
            <button
              class="joinBtn"
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</div>
`;
